### PR TITLE
USHIFT-673: Clean up config file format

### PR DESCRIPTION
**Pull request by fzdarsky**
- Groups cluster-, node-, and apiserver-related params.
- Wires-through the cluster-name parameter.
- Updates /etc/microshift/config.yaml.default to reflect the format correctly.
- Removes 'url' field from being user-configurable.
- Fixes a couple of parameter descriptions.

Closes [USHIFT-673](https://issues.redhat.com//browse/USHIFT-673)

### CODE REVIEW

**Comment by mangelajo** [comment]
For me this structure makes a lot of sense. I believe it would be a good idea to go with this structure for 4.12

**Comment by mangelajo** [comment]
hostnameOverride: "" is a better name than what we had before because it's more specific about the intent, and seems ok to leave it empty, while before it was confusing if you had to provide it.

**Review by mangelajo — APPROVED** [review]

File: pkg/config/config.go [hunk 1]
```diff
@@ -188,10 +206,11 @@ func NewMicroshiftConfig() *MicroshiftConfig {
 	}
 
 	return &MicroshiftConfig{
-		LogVLevel:       0,
+		LogVLevel:       2,
```
[the remark this answers is not in the document]
Comment (dinhxuanvu) [reply pkg/config/config.go:209]: I think this is because the new default LogLevel is "Normal" which is equivalent to "2" not "0". This is the standard verbosity setting that is currently used on [KAS operator](https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/targetconfigcontroller/targetconfigcontroller.go#L492) as well.

File: test/config_bad_subjectaltnames.yaml [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
+  clusterNetwork:
+    - cidr: '10.20.30.40/16'
+  serviceNetwork:
+    - '40.30.20.10/16'
   serviceNodePortRange: 30000-32767
+node:
+  hostnameOverride: node1
+  nodeIP: '1.2.3.4'
+apiServer:
+  subjectAltNames:
```
Comment (dinhxuanvu) [inline test/config_bad_subjectaltnames.yaml:15]: I'm not gonna lie and say that the variable name as "subjectAltNames" indicates some sort of string name but the value is used as networking/IP info is just confusing. However, it is what it is so I make my peace with that.

File: docs/howto_config.md [hunk 3]
```diff
@@ -10,19 +10,21 @@ The MicroShift configuration file must be located at `~/.microshift/config.yaml`
...
 network:
   clusterNetwork:
     - cidr: ""
   serviceNetwork:
     - ""
   serviceNodePortRange: ""
-dns:
-  baseDomain: ""
+node:
+  hostnameOverride: ""
```
[the remark this answers is not in the document]
Comment (mangelajo) [reply docs/howto_config.md:23]: I believe @fzdarsky picked this one to align with the kubelet parameter of the same meaning. It's long and not nice, but descriptive of the purpose.

I am ok with both options though.

File: test/config_bad_subjectaltnames.yaml (hunk 2 above)
Comment (mangelajo) [reply test/config_bad_subjectaltnames.yaml:15]: This is an example of "bad subject alt names" (by the file name `config_bad_subjectaltnames.yaml`) So I guess it's some test checking things that we should not use in that field

File: docs/howto_config.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply docs/howto_config.md:23]: Was thinking about just "name", but "hostnameOverride" would have two advantages:
1. It makes explicit what the parameter is eventually used for, namely being passed to the kubelet via the "hostnameOverride" parameter. The kubelet otherwise defaults the node name to host name (like we do).
2. It makes it more intuitive that this param should be left empty unless there's a reason to override the default behavior.

File: pkg/config/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply pkg/config/config.go:209]: Correct. It also fixes a problem that without `config.yaml` present, we've been defaulting to 0 and with `config.yaml` present, to 2 (https://github.com/openshift/microshift/blob/main/pkg/config/config.go#L329 and https://github.com/openshift/microshift/blob/main/pkg/config/config.go#L533-L534).

File: test/config_bad_subjectaltnames.yaml (hunk 2 above)
Comment (fzdarsky) [reply test/config_bad_subjectaltnames.yaml:15]: One would expecdt subjectAltNames to be FQDNs, I agree. Seeing IPs here also only makes sense to me as this is for a unit test for a failure case.

File: docs/howto_config.md [hunk 4]
```diff
@@ -10,19 +10,21 @@ The MicroShift configuration file must be located at `~/.microshift/config.yaml`
...
+  nodeIP: ""
```
Comment (benluddy) [inline docs/howto_config.md:24]: This field is (currently) used for many things that are unrelated to Node/kubelet. Unless that's going to change, I think putting it into a "node" config group is misleading.

File: pkg/cmd/init.go [hunk 5]
```diff
@@ -278,8 +275,8 @@ func certSetup(cfg *config.MicroshiftConfig) (*certchains.CertificateChains, err
 					"openshift.default",
 					"openshift.default.svc",
 					"openshift.default.svc.cluster.local",
-					"api." + config.DefaultClusterName + "." + cfg.BaseDomain,     // TODO: OpenShift actually uses  api.$ClusterName.$BaseDomain
-					"api-int." + config.DefaultClusterName + "." + cfg.BaseDomain, // TODO: OpenShift actually uses  api.$ClusterName.$BaseDomain
+					"api." + cfg.ClusterName + "." + cfg.BaseDomain,
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply pkg/cmd/init.go:278]: According to [OCP documentation](https://docs.openshift.com/container-platform/4.11/installing/installing_platform_agnostic/installing-platform-agnostic.html#installation-dns-user-infra_installing-platform-agnostic), all DNS records are subdomains of `<cluster_name>.<base_domain>`:
```
api.<cluster_name>.<base_domain>
api-int.<cluster_name>.<base_domain>
*.apps.<cluster_name>.<base_domain>
```
And the `openshift-installer`'s [`install-config.yaml`](https://docs.openshift.com/container-platform/4.11/installing/installing_platform_agnostic/installing-platform-agnostic.html#installation-bare-metal-config-yaml_installing-platform-agnostic) takes them in via fields `metadata.name` and `baseDomain`.

In OCP, of course, this is because the cluster-name receives a random suffix to avoid failures when clusters get deleted and re-created under the same name. In MicroShift that won't be the case.

So there isn't really a utility and we could redefine `baseDomain` to also include the cluster name, but my understanding was we wanted to align.

File: docs/howto_config.md [hunk 6]
```diff
@@ -10,19 +10,21 @@ The MicroShift configuration file must be located at `~/.microshift/config.yaml`
...
 
 ```yaml
-subjectAltNames:
-  - ""
-nodeName: ""
-nodeIP: ""
-url: ""
+cluster:
+  name: ""
+  baseDomain: ""
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply docs/howto_config.md:15]: IDK. The definition you've linked seems to differ, but that may just be an inconsistency with the documentation/installer (see below).

File: docs/howto_config.md (hunk 4 above)
Comment (fzdarsky) [reply docs/howto_config.md:24]: I can see your point, though I wouldn't call those "many things" unrelated:
We currently use nodeName for a) TLS certs for services on the node, b) in etcd config making the node member, c) as name for the node in kubelet, d) in mDNS to announce the node. In all cases but kubelet, that node name will be a domain name that resolves to the same IP address. In the case of kubelet, the node name _defaults_ to the host name, thus also pointing to the same IP address. It _may_ be overridden (with `--override-hostname`) to something else, of course, but would still need to be resolvable.

Is there another grouping that'd make more sense to you? Should I revert to "nodeName" (although that also suggests "node") or use `hostnameOverride` (although we use it for things unrelated to kubelet)?

Comment (fzdarsky) [reply docs/howto_config.md:24]: Maybe. @pacevedom WDYT?

File: docs/howto_config.md [hunk 7]
```diff
@@ -10,19 +10,21 @@ The MicroShift configuration file must be located at `~/.microshift/config.yaml`
...
   serviceNetwork:
     - ""
   serviceNodePortRange: ""
-dns:
-  baseDomain: ""
+node:
+  hostnameOverride: ""
+  nodeIP: ""
+apiServer:
+  subjectAltNames:
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply docs/howto_config.md:26]: Maybe. @pacevedom WDYT?

[the remark this answers is not in the document]
Comment (pacevedom) [reply docs/howto_config.md:26]: Not sure. The purpose of the apiserver config is to tie certificates to hostnames, which is different. This one is about adding names to the external serving certificate.

File: docs/howto_config.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (dhellmann) [reply docs/howto_config.md:23]: OK, I think that makes sense.

File: pkg/config/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhellmann) [reply pkg/config/config.go:209]: OK

File: pkg/cmd/init.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (benluddy) [reply pkg/cmd/init.go:278]: I see, so the `spec.baseDomain` field of the config/v1 DNS type is constructed _by the installer_ by concatenating cluster name and the "baseDomain" specified in the install config:

https://github.com/openshift/installer/blob/780b29adbe5ba47a1b41ffeade875583282995db/pkg/asset/manifests/dns.go#L86
https://github.com/openshift/installer/blob/780b29adbe5ba47a1b41ffeade875583282995db/pkg/types/installconfig.go#L196

So, to me, this means that ignoring cluster name entirely would be the choice most consistent with OCP's config/v1 API, perhaps with the sample value for MicroShift's `dns.baseDomain` field being "microshift.example.com".

File: test/config_bad_subjectaltnames.yaml (hunk 2 above)
Comment (fzdarsky) [reply test/config_bad_subjectaltnames.yaml:15]: As SAN is clearly defined in the technical standard, using that name as param is the most descriptive IMO. It's just my ignorance that I didn't know the standard allowed IPs :)

I wouldn't do anything with the YAML. It's a unit test case, after all, not user documentation.

File: pkg/cmd/init.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mangelajo) [reply pkg/cmd/init.go:278]: Based on this conversation, I tend to think that we may want to drop the clustername altogether , and just keep the base domain. As the clustername is just a construct of the installer then injected into the cluster itself.

This would leave customers freedom to deploy the way they want.

File: docs/howto_config.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply docs/howto_config.md:14]: Reverted this.

File: pkg/cmd/init.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply pkg/cmd/init.go]: Reverted this.

File: pkg/controllers/kube-controller-manager_test.go [hunk 8]
```diff
@@ -67,7 +67,7 @@ func TestConfigure(t *testing.T) {
 		"--secure-port=10257",
 		fmt.Sprintf("--service-account-private-key-file=%s", kcmServiceAccountPrivateKeyFile()),
 		"--use-service-account-credentials=true",
-		"-v=0",
+		"-v=2",
```
[the remark this answers is not in the document]
Comment (fzdarsky) [reply pkg/controllers/kube-controller-manager_test.go:70]: "Normal" corresponds to "2", correct?

**Review by mangelajo — APPROVED** [review]
/approve

File: docs/howto_config.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (oglok) [reply docs/howto_config.md:22]: I'm up for `hostnameOverride` too. It's a well-known param for kubelet.

File: pkg/controllers/kube-controller-manager_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (oglok) [reply pkg/controllers/kube-controller-manager_test.go:70]: @dgrisonnet we commented about that [here](https://github.com/openshift/microshift/pull/1187/commits/d30759b34b3cbbc2a8d8d1b4ecec7e6d1efb2417#r1045846861).  Do you mean that?

**Comment by oglok** [comment]
I'll take the shot of tagging this. I think there were great points and discussions on the PR. Some of them have been addressed and I think the result makes a lot of sense. If there is a need to make a minor change, it could be done in a follow PR, but let's get this merged since it's important for 4.12.

/lgtm
